libmobilecoin/src/chacha20_rng.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::common::*;
+use mc_util_ffi::*;
+use rand_chacha::ChaCha20Rng;
+use rand_core::{RngCore, SeedableRng};
+use std::{convert::TryInto, sync::Mutex};
+
+pub type McChaCha20Rng = ChaCha20Rng;
+
+// McU128 facilitates conversion between u128 and 16 bytes of u8
+pub struct McU128 {
+    pub bytes: [u8; 16],
+}
+
+impl McU128 {
+    pub fn from_u128(val: u128) -> McU128 {
+        McU128 {
+            bytes: val.to_be_bytes(),
+        }
+    }
+
+    pub fn to_u128(&self) -> u128 {
+        u128::from_be_bytes(self.bytes)
+    }
+}
+
+impl IntoFfi<McU128> for McU128 {
+    #[inline]
+    fn error_value() -> McU128 {
+        McU128 {
+            bytes: [u8::MAX; 16],
+        }
+    }
+
+    #[inline]
+    fn into_ffi(self) -> McU128 {
+        self
+    }
+}
+
+impl_into_ffi!(FfiOwnedPtr<McU128>);
+impl_into_ffi!(Mutex<McChaCha20Rng>);
+
+/// Returns a new ChaCha20Rng instance initialized with the
+/// seed value provided by the u64 long_val parameter
+///
+/// # Arguments
+///
+/// * `long_val` - an unsigned 64 bit value to use as the rng seed
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_create_with_long(
+    long_val: u64,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> FfiOptOwnedPtr<Mutex<McChaCha20Rng>> {
+    ffi_boundary_with_error(out_error, || {
+        Ok(Mutex::new(McChaCha20Rng::seed_from_u64(long_val)))
+    })
+}
+
+/// Returns a new ChaCha20Rng instance initialized with the
+/// seed value provided by the bytes data, which must be at
+/// least 32 bytes (only the first 32 bytes will be used)
+///
+/// # Arguments
+///
+/// * `bytes` - 32 bytes of data to use as the rng seed
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+#[no_mangle]
+pub extern "C" fn mc_chacha20_rng_create_with_bytes(
+    bytes: FfiRefPtr<McBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> FfiOptOwnedPtr<Mutex<McChaCha20Rng>> {
+    ffi_boundary_with_error(out_error, || {
+        let bytes: [u8; 32] = bytes
+            .as_slice()
+            .try_into()
+            .expect("seed size must be 32 bytes");
```

Comment:
I'm not sure panicking/`expect` is the right thing to do here. `ffi_boundary_with_error` does handle panics, but I'm not sure what the result would be, let alone whether it's reasonable when a caller passes < 32 bytes

Reply:
updated...